File: backend/helpers/fileConverter.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const convertFile = async (filePath) => {
    return new Promise((resolve, reject) => {
        if (!fs.existsSync(filePath)) {
            return reject(new Error(`Error: File not found - ${filePath}`));
        }

        const outputFilePath = filePath.replace(/\.\w+$/, '.pdf');

        //Run LibreOffice CLI command
        const command = `soffice --headless --convert-to pdf "${filePath}" --outdir "${path.dirname(filePath)}"`;

        exec(command, (error, stdout, stderr) => {
            if (error) {
                console.error('LibreOffice conversion error:', stderr);
                return reject(error);
            }

            if (!fs.existsSync(outputFilePath)) {
                return reject(new Error("Conversion failed: Output PDF not found."));
            }

            console.log(`Successfully converted: ${outputFilePath}`);
            resolve(outputFilePath);
        });
    });
};


module.exports = convertFile;

File: backend/app/helpers/file_converter.py
import os
import subprocess
from pathlib import Path


def convert_file(file_path: str) -> str:
    input_path = Path(file_path)

    if not input_path.exists():
        raise FileNotFoundError(f"Error: File not found - {file_path}")

    output_file_path = str(input_path.with_suffix(".pdf"))

    command = [
        "soffice",
        "--headless",
        "--convert-to",
        "pdf",
        str(input_path),
        "--outdir",
        str(input_path.parent),
    ]

    result = subprocess.run(
        command,
        capture_output=True,
        text=True,
    )

    if result.returncode != 0:
        raise RuntimeError(
            f"LibreOffice conversion error: {result.stderr or result.stdout}"
        )

    if not os.path.exists(output_file_path):
        raise RuntimeError("Conversion failed: Output PDF not found.")

    print(f"Successfully converted: {output_file_path}")
    return output_file_path